crates/ide-db/src/imports/import_assets.rs

```diff
@@ -536,6 +537,29 @@ fn trait_applicable_items(
         required_assoc_items.insert(assoc);
         Some(assoc_item_trait.into())
     })
+    .filter(|candidate_trait_id| {
+        // we care about the following cases:
+        // 1. Trait's definition crate
+        // 2. Definition crates for all trait's generic arguments
+        //     a. This is recursive for fundamental types: `Into<Box<A>> for ()`` is OK, but `Into<Vec<A>> for ()`` is *not*.
+        // 3. Receiver type definition crate
+        //    a. This is recursive for fundamental types
+        let candidate_trait_id: TraitId = *candidate_trait_id;
+        let defining_crate_for_trait = Trait::from(candidate_trait_id).krate(db);
+        let Some(receiver_fingerprint) = trait_candidate.receiver_ty.fingerprint_for_trait_impl()
+        else {
+            return false;
+        };
+        let definitions_exist_in_trait_crate = db
+            .trait_impls_in_crate(defining_crate_for_trait.into())
+            .has_impls_for_trait_and_self_ty(candidate_trait_id, receiver_fingerprint);
+
+        let definitions_exist_in_receiver_crate = db
+            .trait_impls_in_crate(trait_candidate.receiver_ty.krate(db).into())
+            .has_impls_for_trait_and_self_ty(candidate_trait_id, receiver_fingerprint);
+
+        definitions_exist_in_trait_crate || definitions_exist_in_receiver_crate
```

Comment:
I doubt the compiler will optimize this to be lazy due to db accesses, so making the latter a closure to be called in the or expression would a bit better perf wise

Reply:
oh, good point. will fix.

Reply:
done!